homeassistant/components/tradfri.py

```diff
@@ -58,25 +57,40 @@ def configuration_callback(callback_data):
         """Handle the submitted configuration."""
         try:
             from pytradfri.api.aiocoap_api import APIFactory
+            from pytradfri import RequestError
+            from uuid import uuid4
         except ImportError:
             _LOGGER.exception("Looks like something isn't installed!")
             return
 
-        api_factory = APIFactory(host, psk_id=GATEWAY_IDENTITY)
-        psk = yield from api_factory.generate_psk(callback_data.get('key'))
-        res = yield from _setup_gateway(hass, config, host, psk,
+        identity = uuid4().hex
+        security_code = callback_data.get('security_code')
+
+        api_factory = APIFactory(host, psk_id=identity, loop=hass.loop)
+        # Need To Fix: currently entering a wrong security code sends
+        # pytradfri aiocoap API into an endless loop.
+        # Should just raise a requestError or something.
+        try:
+            key = yield from api_factory.generate_psk(security_code)
+        except RequestError:
+            configurator.async_notify_errors(hass, instance,
+                                             "Security Code not accepted.")
+            return
+
+        res = yield from _setup_gateway(hass, config, host, identity, key,
                                         DEFAULT_ALLOW_TRADFRI_GROUPS)
 
         if not res:
-            hass.async_add_job(configurator.notify_errors, instance,
-                               "Unable to connect.")
+            configurator.async_notify_errors(hass, instance,
+                                             "Unable to connect.")
             return
 
         def success():
             """Set up was successful."""
-            conf = _read_config(hass)
-            conf[host] = {'key': psk}
-            _write_config(hass, conf)
+            conf = load_json(hass.config.path(CONFIG_FILE))
+            conf[host] = {'identity': identity,
+                          'key': key}
+            save_json(hass.config.path(CONFIG_FILE), conf)
             hass.async_add_job(configurator.request_done, instance)
```

Comment:
No, this is still wrong. This function starts with `async_` so guess what, it's async 😉 

This context is not decorated with `@asyncio.coroutine` or `@callback`, so you are not allowed to call async methods.

Reply:
hehe, I see. I just reverted my change. But I guess it was wrong all along.